decisions: new categories

markus2330 opened the pull request:
## Basics

These points need to be fulfilled for every PR:

- [ ] Short descriptions of your changes are in the release notes
      (added as entry in `doc/news/_preparation_next_release.md` which
      contains `_(my name)_`)
      **Please always add something to the release notes.**
- [ ] Details of what you changed are in commit messages
      (first line should have `module: short statement` syntax)
- [ ] References to issues, e.g. `close #X`, are in the commit messages.
- [ ] The buildservers are happy. If not, fix **in this order**:
  - [ ] add a line in `doc/news/_preparation_next_release.md`
  - [ ] reformat the code with `scripts/dev/reformat-all`
  - [ ] make all unit tests pass
  - [ ] fix all memleaks
- [ ] The PR is rebased with current master.

If you have any troubles fulfilling these criteria, please write
about the trouble as comment in the PR. We will help you.
But we cannot accept PRs that do not fulfill the basics.

## Checklist

Check relevant points but **please do not remove entries**.
For docu fixes, spell checking, and similar none of these points below
need to be checked.

- [ ] I added unit tests for my code
- [ ] I fully described what my PR does in the documentation
      (not in the PR description)
- [ ] I fixed all affected documentation
- [ ] I added code comments, logging, and assertions as appropriate (see [Coding Guidelines](https://master.libelektra.org/doc/CODING.md))
- [ ] I updated all meta data (e.g. README.md of plugins and [METADATA.ini](https://master.libelektra.org/doc/METADATA.ini))
- [ ] I mentioned every code not directly written by me in [THIRD-PARTY-LICENSES](doc/THIRD-PARTY-LICENSES)

## Review

Reviewers will usually check the following:

- [ ] Documentation is introductory, concise, good to read and describes everything what the PR does
- [ ] Examples are well chosen and understandable
- [ ] Code is conforming to [our Coding Guidelines](https://master.libelektra.org/doc/CODING.md)
- [ ] APIs are conforming to [our Design Guidelines](https://master.libelektra.org/doc/DESIGN.md)
- [ ] Code is consistent to [our Design Decisions](https://master.libelektra.org/doc/decisions)

## Labels

If you are already Elektra developer:

- Add the "work in progress" label if you do not want the PR to be reviewed yet.
- Add the "ready to merge" label **if the basics are fulfilled** and you also
  say that everything is ready to be merged.

kodebach reviewed (commented):
I looked through these decisions and had a few thoughts. The big one is the escaped key names. I don't really like the "lazy value" solution, but I don't really have an alternative either.

Additionally, I have some concerns that didn't fit anywhere:
1. **Memory layout**: In particular because of `mmapstorage`, but possibly elsewhere too, we basically require that implemenations use a certain memory layout. This limits alternative implementations (e.g. I think in Rust it would be much easier than in C to use a separate struct without a meta-keyset for meta-keys). It is also a bit weird that because of `mmapstorage` the memory layout is basically part of the API, and yet we use an opaque struct (I know it has other reasons like safe locking, but it is still weird).
2. **Key Names**: I think it might actually make sense to create a separate `struct KeyName`. I still need to think about how this would work exactly, but there are a few reasons why I think this could be good:
   - Often a `Key *` argument is used when you just need a key name. This is because with a `Key *` we know the name is valid and we get an unescaped name. Using a `Key *` here makes the API a bit confusing.
   - There could be a richer API for manipulating key names without relying on the escaped name (e.g. concatenating two full key names). With the current situation, all these functions would need to be part of the API for a key. Adding such functions to the key API is certainly not minimal.
   - If the struct is not opaque, stack instances can be created. This means we can get rid of `ksLookupByName`. However, then we get problems with locking key names so maybe this part is not a good idea.
   - This might also make it possible to have a `keyNew` without using an escaped name and without using lots of macro tricks and array arguments to make the API somewhat usable.

markus2330 commented on doc/decisions/binary.md:
```diff
@@ -0,0 +1,40 @@
+
+- is the only metadata that allows more values (instead of making the validation stricter)
+- is the only API that modifies metadata
+
```
Can you please explain the inconsistencies in more detail please? My proposal is to not let the API modify any metadata, why would it affect `KEY_FLAG_RO_META`?

kodebach replied:
Your proposal does affect `KEY_FLAG_RO_META` in a positive way. The issue is with the current implementation, hence the note in the "Problems" section.

The name `KEY_FLAG_RO_VALUE` indicates that this flag determines whether a key's value is writable or not. However, this is not entirely true. For binary values `KEY_FLAG_RO_META` is also relevant, because we must modify the `meta:/binary` metadata.

The proposed solution doesn't have the problem, because metadata is not used anymore.

markus2330 commented on doc/decisions/binary.md:
```diff
@@ -0,0 +1,40 @@
+- make keys with untyped value per default (no assumption about type, i.e., byte-array)
+- all current types are not binary, so type = string is the way to indicate a string doesn't
+  - contain null bytes
+  - is not null (not only indicator of structure without value)
```
Where do we need this requirement? Is this a CORBA string defined like this?

kodebach replied:
> - Characters, as defined in ISO Latin-1 (8859.1) and other single- or multi-byte character sets
> - [...]
> - A string type, which consists of a variable-length array of characters; the length of the string is a non-negative integer, and is available at run-time. The length may have a maximum bound defined.
>
> ~ CORBA Specification Version 3.4, Part 1, Section 5.2.4.1 Basic Types

So yes, CORBA (kind of) defines this. However, I really wouldn't use this definition directly. The definition for "string" is fine, but relies on the definition of a "character", which is incredibly vague, and almost useless. The way I read this, it's not even necessary that all characters use the same encoding.

The actual reason, why I suggested this addition is that without an underlying text encoding a string is hardly a string at all. The encoding doesn't have to be stored, it could be contextual knowledge the application has, be derived from the metadata, etc. The encoding could even be different for each key. The only thing that matters IMO is that on somewhere, someone agreed on a way of turning the byte sequence into text. Otherwise, whats the point...

> Where do we need this requirement?

For example in `kdb get /some/key`. We print the key value and assume the terminal is configured with the correct text encoding. We don't care what this encoding actually is, but it has to exist, otherwise we will always print garbage.

markus2330 commented on doc/decisions/simplify_api.md:
```diff
@@ -0,0 +1,94 @@
+- keySetString
+- keyGetBinary
+- keySetBinary
+- ksClear
+- ksCut
```
To keep @lawli3t and your work in the defined scope we must, as far as possible, only talk about reduction of API.

I think all of the mentioned functions can be implemented outside of core, so simply removing ksCut should be the obvious choice?

Btw. there is also the idea of ksForEach and ksFilter, see tests/abi/testabi_ks.c which also let you easily implement ksCut.

kodebach replied:
> I think all of the mentioned functions can be implemented outside of core

In theory yes, but this would a bad idea.

- `ksFindHierarchy`: Since we know have `ksSearch`, this would be the most feasible to move out of the core. However, the optimizations the avoid allocating a new `Key` when searching for `end` are only possible in the core: https://github.com/ElektraInitiative/libelektra/blob/f0b07d837f7e24e55e4a1cbb65f83ea4ac68a151/src/libs/elektra/keyset.c#L1246-L1270
- `ksCopyRange`: If this was done outside the core, you need to do a separate `memcpy` for every `Key *`. Inside the core, we can just use a single `memcpy` to copy the whole section of the `KeySet`. 
- `ksRemoveRange`: Here the multiple `memcpy` issue is even worse, because we need to always "close the gap" in the array of the `KeySet`. For example, if we have a keyset with 1000 Keys and want to remove the first 900 of them, this would result in 900 `memmove` moving 900, then 899, 898, ... Keys at a time. The optimized version doesn't need a memmove at all, since we are removing until the end, but even if we kept the last key, it would be a single `memmove` of 899 keys instead of 899 separate calls.

> Btw. there is also the idea of ksForEach and ksFilter, see tests/abi/testabi_ks.c which also let you easily implement ksCut.

`ksForEach` is pretty useless now that we can simple use an indexed `for` loop. `ksFilter` could still be useful, but has several problems:
- It uses `keyDup`, so the behaviour is different.
- C doesn't support lambdas so using local data is hard in general, and impossible in this case, because the don't have an additional `void *` argument.
- It has the same multiple `memcpy` instead of a single one problem that is mentioned above.
- It cannot be used with the optimized search for the end key, that `ksFindHierarchy` and `ksCut` use. You would have to call `ksFindHierarchy` first and then pass the result to the filter, but is not possible because of the lambda issue.

> To keep @lawli3t and your work in the defined scope we must, as far as possible, only talk about reduction of API.

I don't think this change would actually be that much work (probably comparable to removing `ksCut`).

I'm not sure what the best way forward is here. I can't really see a reason why `ksCut` should be in the 1.0 API. Compared to the other functions it is pretty specific. However, we can't just remove it, because there is no good replacement.

There is also another opportunity here:
- `ksFindHierarchy` and `ksSearch` could be combined into
   ```c
   /**
    * searches for root and the hierarchy below root in ks
    * if root is in ks:
    *    - sets start to the position of root
    *    - sets end to the position of the first key that is not equal to or below root
    *    - returns true
    * if root is not in ks, but ks contains a key that is below root:
    *    - sets start to the position where root would be inserted, i.e. the first key inside ks that is sorted after root
    *    - sets end to the same value as start
    *    - returns false
    * if root is not in ks, and ks contains no key that is below root:
    *    - sets start to ksGetSize (ks)
    *    - sets end to ksGetSize (ks)
    *    - returns false
    */
   bool ksFind (const KeySet * ks, const Key * root, elektraCursor * start, elektraCursor * end);

   // shouldn't actually be part of the core API, but a helper for C only
   static inline elektraCursor ksFindHierarchy (const KeySet * ks, const Key * root, elektraCursor * end) {
     elektraCursor start, end;
     bool found = ksFind (ks, root, &start, &end);
     if (found) {
       return start;
     }
     if (end != NULL) {
       *end = ksGetSize (ks);
     }
     return ksGetSize (start);
   }

   // shouldn't actually be part of the core API, just here to show the replacement
   static inline ssize_t ksSearch (const KeySet * ks, const Key * key) {
     elektraCursor start;
     bool found = ksFind (ks, root, &start, NULL);
     return found ? start : -start - 1;
   }
   ```
   In fact, if properly optimized, this `ksFind` could be used for `ksLookup` with no options.
- `ksRemoveRange` can be merged into `ksCopyRange`:
  ```c
  // new, copies keys in range start <= index < end from source to dest
  // if remove == true, the copied keys are also removed from source
  // if dest == NULL, no copying happens, i.e. removed keys are deleted
  // use ksCopyRange (ks, start, -1) to copy all keys from start to the end of ks
  // returns the number of copied/removed keys
  ssize_t ksCopyRange (KeySet * dest, KeySet * source, elektraCursor start, elektraCursor end, bool remove);

  static inline ssize_t ksRemoveRange (KeySet * ks, elektraCursor start, elektraCursor end) {
    return ksCopyRange (NULL, ks, start, end, true);
  }
  ```

All in all, however, I think it is probably best to keep things as they are for the time being and maybe make `ksCut` internal (removing is probably not feasible).

markus2330 commented on doc/decisions/simplify_api.md:
```diff
@@ -0,0 +1,94 @@
+## Problem
+
+According to src/libs/elektra/symbols.map the public core API has 124 functions, which
+is arguable too much for a key-value API which has as highest goal simplicity.
```
> By my count

Thank you, I added this information in 4781dcfb9437a18294de045ecd4791ade8e177d9

> As a sidenote, we should consider creating separate folders and symbols files for the libelektra-core and libelektra-kdb libraries. The current situation can be a bit confusing.

This is actually quite important: #4219 (otherwise we would get quite a mess when the rust code gets added)

markus2330 commented:
@kodebach thank you very much for the review! I updated the decisions.

> 1. Memory layout

I added memory_layout.md

> 2. Key Names

I added key_name.md

kodebach commented on doc/decisions/memory_layout.md:
```diff
@@ -0,0 +1,32 @@
+
+## Decision
+
+Have versioning and memory layout checks of cache files written by `mmapstorage`.
```
Unless I'm missing something important, this cannot work.

1. `mmapstorage` is written C, so it itself can only create `Key`s/`KeySet`s using the memory layout of the C implementation.
2. The magic numbers `mmapstorage` uses only guard against byte-order changes, e.g. if the cache was copied between machines with different architectures.
3. If you give `mmapstorage` a `KeySet *` that uses a different memory layout, it will happily try and interpret that as a `struct _KeySet *` using the C memory layout. With a lot of luck it might even succeed in storing _something_ into the cache.
4. If a cache file was written by `mmapstorage`, it will be able to read it on the same architecture. It will happily treat the data from the cache as is if it followed the C memory layout. That means it will read it's own magic data and verify that is correct. This will work, because `mmapstorage` always uses the C memory layout. Then it is going to update the pointers. Or at least it pretends to, because where `mmapstorage` thinks the pointers are is probably not where the pointers are in the completely different memory layout. Finally, it will assign some pointer to where it thinks the KeySet array is, but probably it will just write garbage to a random memory location, or even segfault.

The only way this can work AFAIK is one of:
1. **ALL** implementation use the exact same memory layout (down to the order of bits in the flags field) for `Key`  and `KeySet`
2. There is a separate `mmapstorage` implementation for every memory layout. Alternatively, there could be a `mmap` support library for every memory layout that holds all the functions that assume a certain memory layout

There would be a bit more flexibility, if we didn't use pointers, but that is not really an option.

kodebach reviewed (commented):
The files for memory layout and key names both kinda miss the point, IMO. The suggestion for `mmapstorage` simply doesn't work and for key names you just defined that the problem is not a problem...

markus2330 commented on doc/decisions/key_name.md:
```diff
@@ -0,0 +1,39 @@
+
+## Decision
+
+A minimal `Key` only consists of a key name (value and meta data are null pointers).
```
I didn't write the rationale yet: I added it now: 1f42496e3865d7bf93963ed1984e160484863823

markus2330 commented on doc/decisions/key_name.md:
```diff
@@ -0,0 +1,39 @@
+
+## Assumptions
+
+- the key name is actually, in every implementation, a plain string
```
Good question, you somewhere assumed there could be a structure needed.

kodebach replied:
I don't think a structure is _needed_, but it is a trade-off. The plain string we have right now allows for very simple and quick sorting of keys. It is also fairly memory efficient. I don't think you could use any less memory without sharing memory between keys.

A more structured key name, or additional stored information (e.g. pointers to each of the parts) could (in theory) speed up certain operations. But like I said, this would mean increased memory cost.

I just think, that this assumption is not actually required for anything that is decided here. Therefore, we shouldn't make the assumption and leave the decision up to the implementation (memory layout problems notwithstanding)

markus2330 replied:
> The plain string we have right now allows for very simple and quick sorting of keys

Thank you for reminding a reason that will probably convince you: The memcmp we use for looking up keys is order of magnitudes faster then anything what we could have with a more structured key name.

I think we should forget about this idea of structuring the name and use unescaped names as much as possible for fast creation of keys.

kodebach replied:
Yes, in the current situation (only C and Rust implementations) the flat byte buffer is definitely the best solution for the key name.

markus2330 commented:
> If you give mmapstorage a KeySet * that uses a different memory layout, it will happily try and interpret that as a struct _KeySet * using the C memory layout. With a lot of luck it might even succeed in storing something into the cache.

afaik this is not true, I hope I clarified it in 8d461967898995db28cd70e06ab3e75ac57550c5

If your concern is if mmapstorage will work with Rust: good question, let us talk about this today

kodebach commented:
> afaik this is not true, I hope I clarified it in [8d46196](https://github.com/ElektraInitiative/libelektra/commit/8d461967898995db28cd70e06ab3e75ac57550c5)

I know about the magic data stuff. However, this only guards against byte-order changes. It cannot detect general memory layout changes.

Let's suppose we have `libelektra-core-A` and `libelektra-core-B`, both implement the core API but they use different versions of `struct _Key`:

```c
// libelektra-core-A
struct _Key
{
  void * data;
  KeySet * meta;
// [...]
};

// libelektra-core-B
struct _Key
{
  KeySet * meta;
  void * data;
// [...]
};
```

`mmapstorage` uses this struct definition directly. In our example, we'll assume `mmapstorage` uses the first definition, because it was built for `libelektra-core-A`. However, our application uses and links against `libelektra-core-B`. For simplicity, we also assume that the cache already exists on disk (e.g. from a `kdb get`) and it was created entirely using `libelektra-core-A`.

Now we run our application. `mmapstorage` uses the first struct. Checks the magic data, which works fine, because nothing has changed. So it goes on and takes the mmap data and creates a KeySet and all the Keys with this data. This also works fine. The problem, however, is that `mmapstorage` of course only knows about the first struct definition. Therefore, it creates the KeySet/Key according to this definition.

The problem happens, when `mmapstorage` returns and our application accesses a `Key` using `libelektra-core-B`. Because `libelektra-core-B` assumes a different struct definition, we suddenly try to interpret the key value as its metadata `KeySet` and all other kinds of weirdness.

In fact to understand this problem, you don't even need to think about `mmap` (`mmap` just makes it much harder to avoid the problem). The same kind of issue would arise, if `struct _Key` was not opaque and we would change its fields. Suddenly an application built for an old version would read the struct completely wrong. This is the reason why `struct _Key` is opaque in the first place.

The core issue here is, that `mmapstorage` _does not_ use functions like `keySetName` or `keySetValue`. Instead it uses the struct definition directly, and therefore directly writes to a certain memory offset.

> If your concern is if mmapstorage will work with Rust: good question, let us talk about this today

Rust doesn't really come into play here. You can have the same issue with two separate C implementations. AFAIK it should be possible to recreate any memory layout in Rust.

kodebach commented on doc/decisions/library_directory_structure.md:
```diff
@@ -0,0 +1,22 @@
+## Problem
+
+Currently src/libs/elektra contains the source of several libraries. Each of the libraries should have its own folder.
+Furthermore, we need to allow rust source to exist in parallel.
```
I would suggest this structure for the repo (formatted as YAML for highlighting):

```yaml
doc:
examples:
scripts:
src:
 c:
  include: = /src/include (only headers that simply combine other headers)
  libs: = /src/libs (one folder per lib, contains headers and source, only libs written in C)
  plugins: = /src/plugins (one folder per plugin, only plugins written in C)
  tools: = /src/tools (one folder per tool, only tools written in C)
  tests: = /tests (only tests written in C)
 rust:
  binding: = /src/bindings/rust
  libs:
   core: = new Rust alternative of libelektra-core
  _: {other folders depend on Rust's needs, AFAIK tests are normally in the same file as source}
 cpp:
  binding: = /src/bindings/cpp (without the tests)
  libs: (one folder per lib, contains headers and source, only libs written in C++)
  plugins: (one folder per plugin, only plugins written in C++)
  tools: (one folder per tool, only tools written in C++)
  tests: (only tests written in C++)
 _: {folders for other languages, "binding" is always the folder for the binding to the C API, rest depends on the language's needs}
```

That way we can also confine the CMake checks for every language (compiler installed, etc.) to `src/<lang>/CMakeLists.txt` folder.

PS. IMO `benchmarks` should be moved into a separate repo. They are not really relevant to users of Elektra and there is no real technical reason to keep them in this repo. In fact, since benchmark results will only be relevant for a specific version of Elektra, we could use one branch per benchmark and use git submodule pointing to the relevant commit of `libelektra`.

kodebach commented on doc/decisions/library_split.md:
```diff
@@ -0,0 +1,40 @@
+
+## Decision
+
+Also allow `libelektra-operations` library to access private Key/KeySet.
+Put struct definitions of Key/KeySet in a separate header file, which gets
+included by parts that need it
```
Continuing from my suggestions above, I would say:
1. It doesn't have to be a single `libelektra-operations` library. If we keep the core minimal, we should probably keep all libraries minimal. Therefore, sometimes two separate libraries make more sense than a single one containing unrelated functions.
2. The header file in question would be `elektra/core-unstable.h`. Anything can include it, but there should be good reason, because it means a breaking change could happen. If a library wants to have a public stable API, but use _unstable_ API, it must mitigate any breaking changes internally.

kodebach replied:
see #4223

kodebach commented on doc/decisions/library_split.md:
```diff
@@ -0,0 +1,40 @@
+
+## Implications
+
+- we need to clearly communicate which plugins/libraries must be exactly in the version of the libelektra-core
```
I would simply say that any guarantees for API/ABI stability, backwards compatibility, etc. only hold, if all `libelektra-*` libraries are built from the same commit of the repository. If you need another configuration for some reason, you're on your own.

For plugins (`libelektraplugin-*`) things are different. Ideally they only use public API, but in practice they probably need to define which versions are compatible. In theory this could also be enforced via the plugin contract.

kodebach replied:
see #4223

kodebach commented:
I know there are already **a lot** of decisions here, but I though of one more thing...

It's basically #3598. The cascading type and the `typedef`s are a nice-to-have, but without the meta type we get a few problems. In fact `Key` also needs to behave differently for `meta:/` names. Most obviously, adding metakeys to `meta:/` keys shouldn't be not allowed. But there is also #3610.

The good news is we don't actually need a new flag anywhere. In `Key` we must sometimes check the current namespace and in `KeySet` we simply check the namespace of the first key. That way any `Key` can be inserted into an empty `KeySet`, but when inserting into non-empty `KeySet`s the new `Key` must be compatible with the existing one(s).

However, the bad news is with the removal of `keyGetMeta`, you cannot check for the existence of metakeys without creating the metadata `KeySet`. One solution is the addition of a `keyHasMeta` function to the new `libelektra-operations` library. The function will access `key->meta` directly and therefore avoid the creation of the `KeySet`.

The problem is that you always need to call `keyHasMeta` first, or accept that the `KeySet` might be created. Another solution would of course be:

```c
// can return NULL
KeySet * keyMeta (const Key * key) {
  return key->meta;
}

// consumes `meta`, don't call `ksDel (meta)` afterwards
// `keyMeta (key) == meta` after calling `keySetMeta (key, meta)`
void keySetMeta (Key * key, KeySet * meta) {
  if (key->meta != NULL) {
    ksDel (key->meta);
  }
  key->meta = meta;
}

// ALTERNATIVELY
// you must call `ksDel (meta)` later
// `keyMeta (key) != meta` after calling `keySetMeta (key, meta)`
void keySetMeta (Key * key, const KeySet * meta) {
  if (key->meta == NULL) {
    key->meta = ksDup (meta);
  } else {
    ksClear (key->meta);
    ksAppend (key->meta, meta);
  }
}
```

This wasn't really an option before, because we kept the old `keySetMeta`. This API would also match `keyName`/`keySetName` and  `keyValue`/`keySetValue`.

markus2330 commented:
> I know there are already a lot of decisions here, but I though of one more thing...

It is a very good thing to think about everything now, so that we have a complete set of decisions. We can anyway decide to postpone or to reject because of time constraints but it makes sense to document everything what we have thought about and discussed.

It would, however, help a lot if you can directly either push to this branch, create suggestions in this PR or also create a new PR. Copying/summarizing&reformatting comments&issues is also quite some work.

(didn't read the other text yet, I hope you bring it to the decision format)

kodebach commented on doc/decisions/simplify_api.md:
```diff
@@ -0,0 +1,94 @@
+- elektraStrLen;
+- elektraStrNCaseCmp;
+- elektraStrNCmp;
+- elektraVFormat;
```
see #4223

kodebach commented on doc/decisions/simplify_api.md:
```diff
@@ -0,0 +1,94 @@
+Unclear:
+
+- keyCmp
+- ksPop
```
see #4223

kodebach commented on doc/decisions/simplify_api.md:
```diff
@@ -0,0 +1,94 @@
+
+- keyCmp
+- ksPop
+- keyNeedSync
```
see #4223

kodebach commented on doc/decisions/simplify_api.md:
```diff
@@ -0,0 +1,94 @@
+- keyCmp
+- ksPop
+- keyNeedSync
+- keyIsBelow
+- keyIsBelowOrSame
+- keyIsDirectlyBelow
```
see #4223

kodebach commented on doc/decisions/elektra_prefix.md:
```diff
@@ -0,0 +1,35 @@
+## Constraints
+
+- In C such libraries, containing the same external identifier, cannot be used together.
+- The [C99 standard, section 5.2.4.1](http://www.open-std.org/jtc1/sc22/wg14/) gives following limit:
+  31 significant initial characters in an external identifier (each universal character name specifying a short identifier of 0000FFFF or less is considered 6 characters, each universal character name specifying a short identifier of 00010000 or more is considered 10 characters, and each extended source character is considered the same number of characters as the corresponding universal character ...
```
see #4223

kodebach commented on doc/decisions/header_file_structure.md:
```diff
@@ -0,0 +1,42 @@
+
+## Problem
+
+kdb.h contains the public API for both libelektra-core and libelektra-kdb.
```
see #4223

kodebach commented on doc/decisions/header_file_structure.md:
```diff
@@ -0,0 +1,42 @@
+2. It contains things that are different levels of "private".
+
+## Constraints
+
```
see #4223

kodebach commented on doc/decisions/header_file_structure.md:
```diff
@@ -0,0 +1,42 @@
+
+## Decision
+
+1.  Some parts are truly private, i.e. shouldn't be used outside the library that defines them. These things are only there because a library was split into multiple .c files. This includes e.g. splitNew and struct \_Split.
+    Here it would be better to use a local header file in the library's source folder.
+    Symbols belonging to this category should not appear at all in the symbols.map file.
+2.  Other things truly private, but must be tested. This includes e.g. most other split* functions or the elektraKeyName* functions.
+    Here a local header could be used as well (requires longer relative includes), or we could create a src/include/testing folder (could be added to include path).
+    Symbols belonging to this category should only appear in the private section of the symbols.map file (we could also create a second private section, if is just there for testing, or adjust some compiler configs for testing).
+3.  Then there are things that I would call internal or pseudo-private. This includes the struct \_Key and struct \_KeySet, but also keyClearSync or elektraMalloc and the high-level functions needed for codegen. These are things that shouldn't be part of the public API, because we don't want to make all the guarantees associated with that, but nonetheless cannot truly be private. In other words,  ...
+    Here I would either create a separate kdbinternal.h or just keep kdbprivate.h, as long as it is a separate file from 4.
+    Symbols belonging to this category should only appear in the private section of the symbols.map file.
+4.  Finally, there is what I'd call unstable or pseudo-public. This includes e.g. ksFindHierarchy or elektraReadArrayNumber. These functions could be public, but for various reasons are not. Maybe they are not well-tested or maybe we just don't want to commit to the function yet. Many of these (but not all) can't be part of libelektra-ease, because the rely on something more private (e.g. directl ...
+    Here I would either create a separate kdbinternal.h or just keep kdbprivate.h, as long as it is a separate file from 3.
+    Symbols belonging to this category should only appear in the private section of the symbols.map file.
```
see #4223

kodebach commented:
I create #4223 with my suggestions.

kodebach commented on doc/decisions/notifications.md:
```diff
@@ -0,0 +1,39 @@
+
+The decision which transport mechanism, e.g. dbus or zeromq, must be
+chosen per computer node or container. Within one Elektra setup,
+they cannot be intermixed (unlike config file formats, where many
```
> Within one Elektra setup, they cannot be intermixed

Is there a technical reason, or would sending notifications via multiple transports at the same time (e.g. both `dbus` and `zeromq`) work, i.e. all messages duplicated to all transports?

If multiple transports is technically possible, we shouldn't use a `symlink` and instead add config keys in `system:/elektra`.

markus2330 replied:
It simply follows the simple rule to keep everything simple as long as nobody needs something more complicated. If someone *really* wants to use several transports on one system, the person can also use some hub that reads from one transport and relays to another. So we do not break any use case by reducing our notification system to one transport.

kodebach replied:
I agree the symlink makes sense to indicate the default (like with resolver and storage). But the decision (at least to me) sounds like we are saying we will never support multiple transports. I still don't see a technical reason why we should restrict ourselves to a single notification transport forever.

I would suggest, we decide that
- the symlink indicates the default notification transport
- we currently only support one transport at a time, so the default is always used
- in future, if the need arises, there may be additional options (e.g. config keys) to enable secondary transports

markus2330 replied:
Yes, of course! The decisions always only describe what we have now or want to implement for 1.0. Decisions can change, especially if assumptions are wrong. That is why it is important to write down assumptions.

So I think this decision describes what we both want.

kodebach replied:
> The decisions always only describe what we have now or want to implement for 1.0. Decisions can change, especially if assumptions are wrong.

But we have to be careful, how we write things. Future developers (maybe we ourselves), might just read the decision and use it as the basis for their own decision. "must be chosen per computer node or container" sounds very much final and could result in APIs or implementation details (in the notification stuff, or elsewhere)  that may be very hard to change without breaking things.

markus2330 replied:
Yes, of course assumptions need to be chosen carefully. They might lead to unsuitability of Elektra in some areas. But as you often stated: we also cannot please everybody, so some assumptions are needed to make Elektra usable.
